src/Strategy/Erlang/ConfigParser.hs

```diff
@@ -0,0 +1,51 @@
+module Strategy.Erlang.ConfigParser (
+  parseConfig,
+  parseErlValue,
+) where
+
+import Data.Functor (void)
+import Data.Text (Text)
+import Data.Void
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+
+type Parser = Parsec Void Text
+
+data ErlValue
+  = Atom Text
+  | Array [ErlValue]
+  | Pair Text ErlValue
+
+parseConfig :: Parser [ErlValue]
+parseConfig = do
+  scn
+  many (parsePair <* char '.')
+
+parseErlValue :: Parser ErlValue
+parseErlValue = parseAtom <|> parseArray <|> parsePair
+
+parseAtom :: Parser ErlValue
+parseAtom = Atom <$> lexeme (takeWhile1P (Just "atom") (== ','))
+
+parseArray :: Parser ErlValue
+parseArray = lexeme (Array <$> between (char '[') (char ']') (many parseErlValue))
+
+parsePair :: Parser ErlValue
+parsePair = do
```

Comment:
this is great, and you probably thought about using the applicative syntax to write this (`<*>`), but realized that you couldn't do so because of the punctuation you want to ignore

this is where a couple of other operators come in:

```hs
(<*) :: Applicative f => f a -> f b -> f a
(<*) = error "TODO: implement me using <$> and <*>"

(*>) :: Applicative f => f a -> f b -> f b
(*>) = error "TODO: implement me using <$> and <*>"
```

the "bird operators" have their beak pointing toward the result we want to keep, discarding the other one. they're left-associative (just like <*> and <$>) -- so you can pretend that there are parentheses wrapping the left side each time: `(((fa <* fb) *> fc) <*> fd)`.

when `f ~ Parser`, we can use these operators to run two parsers in sequence, discarding the result of one of them. for example, here:

```hs
parsePair :: Parser ErlValue
parsePair = pure Pair <* cc "{" <*> parseAtom <* cc "," <*> parseErlValue <* cc '}'
```

going from left-to-right, because the operators are left-associative:
```hs
pure Pair :: Parser (ErlValue -> ErlValue -> ErlValue)

-- ignoring the {, just keeping the Pair constructor we have so far
pure Pair <* cc "{" :: Parser (ErlValue -> ErlValue -> ErlValue)

-- applying the Pair constructor to the result of parseAtom
pure Pair <* cc "{" <*> parseAtom :: Parser (ErlValue -> ErlValue)

-- ignoring the ,
pure Pair <* cc "{" <*> parseAtom <* cc ',' :: Parser (ErlValue -> ErlValue)

-- applying the remaining Pair constructor to the result of parseErlValue
pure Pair <* cc "{" <*> parseAtom <* cc ',' <*> parseErlValue :: Parser ErlValue

-- ignoring the }
pure Pair <* cc "{" <*> parseAtom <* cc ',' <*> parseErlValue <* cc '}' :: Parser ErlValue
```

---

and if that wasn't enough, `pure anything <*>` has a name:

```hs
-- note the similarity between the types of <$ and <*
(<$) :: Functor f => a -> f b -> f a
x <$ f = error "TODO: implement me using <$>"
```

where `<$` again is kind of "pointing at the result we want to keep"

so our final parser:

```hs
parsePair :: Parser ErlValue
parsePair = Pair <$ cc "{" <*> parseAtom <* cc "," <*> parseErlValue <* cc '}'
```

Reply:
What's the benefit of doing this exact parser that way?  At first glance, that's less readable to me, but mostly because of the `<*>` operator.  I think the spacing helps with readability here.  Frankly, though, I could be convinced otherwise.

Reply:
The general benefit of using applicative syntax over monadic syntax is that the applicative syntax is _more restrictive_, so is more obviously-correct at a glance.

For example, if we're binding out values monadically, we can use those values to _change the behavior_ of our computation:

```hs
something = do
  foo <- bar
  case foo of
    ...
```

..so when I see a `do` block, I _have to_ carefully inspect it. I also have to inspect it for possible mistakes with binding variables and using the wrong variables later

With applicative, I can't do that. Applicative only allows us to "thread some structure" through a bunch of computations, without inspecting the values.

You can also play around with the spacing a bit for readability, e.g.,
```hs
parsePair :: Parser ErlValue
parsePair = Pair
  <$ cc "{"
  <*> parseAtom
  <* cc ","
  <*> parseErlValue
  <* cc '}'
```